### PR TITLE
Quadlet - Allow setting Service WorkingDirectory for Kube units

### PR DESCRIPTION
**Pull request by ygalblum**
Add key for Quadlet to set WorkingDirectory to the directory of the YAML or Unit file Add Doc
Add E2E tests
Add System test

#### Does this PR introduce a user-facing change?
Yes

```release-note
Quadlet - Allow the user to instruct Quadlet to set the service working directory relative to the YAML or Unit files
```

Resolves #17177

### CODE REVIEW

**Comment by ygalblum** [comment]
> What am I missing?

The issue is how transferable the unit file is. For example, let's say that we have a service that different users want to run. They can each place it under `/etc/containers/systemd/users/$UID`. Now, assuming they want the path to point into this directory, only the generator can handle it automatically. Your suggestion will require setting this field differently for each user.

**Comment by Luap99** [comment]
Right but for that you can use the systemd specifiers (`%t`,`%h`,etc...)
I just wonder if this is really needed? But if you think that makes in easier to use for users I have no objections.

code LGTM

**Comment by ygalblum** [comment]
Maybe the specifiers can solve the use case I added. But, from the discussion in #17177 it seems that having Quadlet resolve the paths is desired

**Comment by rhatdan** [comment]
I think the issue right now is that this is random, and not something the user can easily expect.

**Comment by Luap99** [comment]
Systemd defaults to the users $HOME when rootless and to `/` when run as root AFAICT.

File: test/e2e/quadlet/workingdir-unit.kube [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## assert-key-is-regex "Service" "WorkingDirectory" "/tmp/podman_.*/quadlet"
```
Comment (edsantiago) [inline test/e2e/quadlet/workingdir-unit.kube:1]: Please do not hardcode `/tmp`.

**Review by edsantiago — CHANGES_REQUESTED** [review]

File: pkg/systemd/quadlet/quadlet.go [hunk 2]
```diff
@@ -1288,3 +1300,38 @@ func handlePodmanArgs(unitFile *parser.UnitFile, groupName string, podman *Podma
 		podman.add(podmanArgs...)
 	}
 }
+
+func handleSetWorkingDirectory(kube, serviceUnitFile *parser.UnitFile) error {
+	// If WoringDirectory is already set in the Service section do not change it
```
Comment (TomSweeneyRedHat) [inline pkg/systemd/quadlet/quadlet.go:1305]:
```suggestion
	// If WorkingDirectory is already set in the Service section, do not change it
```

File: docs/source/markdown/podman-systemd.unit.5.md [hunk 3]
```diff
@@ -525,6 +525,7 @@ Valid options for `[Kube]` are listed below:
 | Network=host                        | --net host                                       |
 | PodmanArgs=\-\-annotation=key=value | --annotation=key=value                           |
 | PublishPort=59-60                   | --publish=59-60                                  |
+| SetWorkingDirectory=yaml            | Set `WorkingDirectory` of unit file to location of the YAML file |
```
Comment (vrothberg) [inline docs/source/markdown/podman-systemd.unit.5.md:528]: Can you reformat the table?

File: docs/source/markdown/podman-systemd.unit.5.md [hunk 4]
```diff
@@ -609,6 +610,12 @@ entry from the unit file takes precedence
 
 This key can be listed multiple times.
 
+### `SetWorkingDirectory=`
+
+Set the `WorkingDirectory` field of the `Service` group of the Systemd service unit file.
+Used to allow `podman kube play` to correctly resolve relative paths.
+Supported values are `yaml` and `unit` to set the working directory to that of the YAML or Quadlet Unit file respectively.
```
Comment (vrothberg) [inline docs/source/markdown/podman-systemd.unit.5.md:617]: What is the default?

Can we also allow for specifying absolute paths?

File: test/e2e/quadlet/workingdir-unit.kube (hunk 1 above)
Comment (ygalblum) [reply test/e2e/quadlet/workingdir-unit.kube]: Right, sorry for that, thanks. Done

File: docs/source/markdown/podman-systemd.unit.5.md (hunk 4 above)
Comment (ygalblum) [reply docs/source/markdown/podman-systemd.unit.5.md:617]: > What is the default?

As far as Quadlet is concerned there is no default. If this key is not set, Quadlet will not do anything

> Can we also allow for specifying absolute paths?

We could. But, that does not provide any advantage over setting `WorkingDirectory` in the `Service` section. The whole point of this key is to have Quadlet calculate the working directory based on the location of the YAML or unit files

Comment (vrothberg) [reply docs/source/markdown/podman-systemd.unit.5.md:617]: Very fair point.  Could you mentioned `WorkingDirectory` in the docs as an alternative in the docs?

Comment (ygalblum) [reply docs/source/markdown/podman-systemd.unit.5.md:617]: I've added a new paragraph about explicitly setting this field

File: test/system/252-quadlet.bats [hunk 5]
```diff
@@ -897,4 +897,62 @@ EOF
...
+Yaml=${yaml_source}
+SetWorkingDirectory=yaml
+EOF
+
+    run_quadlet "$quadlet_file"
+    service_setup $QUADLET_SERVICE_NAME
+
+    # Ensure we have output. Output is synced via sd-notify (socat in Exec)
+    run journalctl "--since=$STARTED_TIME" --unit="$QUADLET_SERVICE_NAME"
+    is "$output" '.*Started.*\.service.*'
```
Comment (edsantiago) [inline test/system/252-quadlet.bats:944]: Not a blocker (right now), but please join the discussion here: https://github.com/containers/podman/pull/19236#pullrequestreview-1534848301

**Review by edsantiago — APPROVED** [review]

**Comment by TomSweeneyRedHat** [comment]
@ygalblum some of the tests are still a bit red.

**Comment by ygalblum** [comment]
> you could kindly update your system test

@edsantiago I've updated my test to not use `journalctl` the same way as in #19284

**Comment by edsantiago** [comment]
Thank you! LGTM but I will let others make the final approval.

**Pull request merged**